Focusing thumb on click or track click

jurajhrib opened the pull request:
Closes #21 

1. Focus thumb when clicked/dragged, stays focused (MouseEvent)
2. Focus single thumb when track clicked, stays focused (MouseEvent)

I've had to introduce ref array for thumbs, which can later be utilized for other purposes as well.
For the first issue, I just removed preventDefault from mouse down handler. I hope it does not break anything.

jurajhrib commented:
The E2E tests are failing, because there is an outline now in the snapshots.

Now, here's the question - should we display the outline by default? With default styles, it can lead to an awkward result when using Labeled range in Chrome. Because some browsers apply outline around all child content, even if everything's absolute. So we get something like this:

<img width="495" alt="Screenshot 2020-01-08 at 18 10 12" src="https://user-images.githubusercontent.com/373788/71999669-22b8fb00-3242-11ea-9e0e-15066a32f019.png">

There are several solutions I can think of:

A. Keep the outline as is (meh)
B. Turn off the default outline in default styles completely (for test and storybook purposes)
C. Refactor the example label so it's outside the thumb element (not sure it will work)

tajo commented:
Could we leverage [focus-visible](https://css-tricks.com/almanac/selectors/f/focus-visible/) to have a nice default outline?

B. is no go because of a11y.

tajo commented:
Also, happy to update the snapshots when this is ready. (it needs to be done on macOS to match the CI environment, I didn't take effort to dockerize this thing yet)

jurajhrib commented:
Ok. So, I've been looking at the specs and how established design-systems do it. Seems like the best practice is, that you do not actually indicate focus on click (it is there, but does not show outline) and you display it when user presses arrow button (or other corresponding keyboard key).

So I'll try to do it this way now, hopefully it will work (and look) as intended.

tajo commented:
I think it's fine to add a polyfill as a dev dep and use it in Storybook. Maybe also add a small note into the README and link it through a code comment? Afaik, there is no other better solution and this API will be eventually the standard.

jurajhrib commented:
I've added the polyfill to stories bootstap file, added the simple css style they recommend and updated the README. The e2e tests seem to be passing now locally, so hopefully CI will also pass.

We're building a design-system with my team and I've decided it's not worth to reinvent the wheel and liked your solution, so we're building our slider on top of yours. That's why I needed the focus behaviour.

jurajhrib commented:
@tajo let me know it it's sufficient

tajo commented:
Hm, how does the focus ring work now? I don't really see a difference now and before. There is no outline when using mouse to drag thumbs. And the style for keyboard focus is the same. Maybe I just don't understand it?

jurajhrib commented:
I see. I will explain the new functionality.

Previously, the thumb did not keep focus on mouse interaction. It lost it immediately due to preventDefault() call. This is now fixed, so the mouse will focus the thumb.

But on mouse interaction, we actually should not indicate the focus visually. That is what `:focus-visible` is for. It let's browser decide, whether the interaction is focus-visual-worthy and applies style if yes.

You can test it by dragging by mouse, and then using arrows. It will display the focus ring.

But the style of the focus ring itself has not changed in this PR, as I did not presume to change default styles, but merely fix the logic.

tajo commented:
> Previously, the thumb did not keep focus on mouse interaction. It lost it immediately due to preventDefault() call. This is now fixed, so the mouse will focus the thumb.

Yep, I got that. That's great. 

I still don't see the styles difference when using mouse though (master version):

![2020-01-10_12-53-28 (1)](https://user-images.githubusercontent.com/1387913/72185824-937a2780-33a8-11ea-9dae-2606db25bfba.gif)

tajo commented:
> Previously, the thumb did not keep focus on mouse interaction. It lost it immediately due to preventDefault() call. This is now fixed, so the mouse will focus the thumb.

Also, can you please add an e2e test for this?

jurajhrib commented:
Here's the behaviour of the new implementation. As you can see - the mouse does not trigger visual focus ring, but keeps the focus on the thumb.  When I then use arrows, the ring appears. That's what most of the UIs are trying to achieve. 

![range](https://user-images.githubusercontent.com/373788/72189639-a79c4400-33fd-11ea-97e2-a4dee8781677.gif)

About the style of the focus ring - I think if we want to change how it looks, there should be a separate PR for that.

jurajhrib commented:
And you are right, this should have an E2E test.

tajo commented:
Oh, I got it now. You don't see the focus ring **on the master** version because the thumb doesn't get properly focused when using mouse (which is the purpose of this PR). Cool! Once there is a test, it's good to go!

tajo commented:
> this should have an E2E test.

Do you need my help with this?

jurajhrib commented:
Sorry for not completing this sooner, I've had limited time to work on OS stuff. I've added one test which checks whether the focus works correctly. The visual behaviour does not need test, as the old tests do the job to test it. 

I hope the test snapshot I've generated on my mac is sufficient for the CI. Hopefully you can now close it (after the tests pass).

tajo commented on e2e/basic.focus.test.ts:
```diff
@@ -0,0 +1,31 @@
+  await page.keyboard.press('ArrowLeft');
+
+  const output = await page.$('#output');
+  expect(await page.evaluate(e => e.textContent, output)).toBe('79.5');
```
My last nit: It would be nice to do this assertion also before `page.keyboard.press('ArrowLeft')` so it's obvious that the value changed and test works.

tajo commented:
> Sorry for not completing this sooner, I've had limited time to work on OS stuff.

No worries. Just want to make sure this nice fix/feature makes it in. Thanks for contributing!

jurajhrib commented:
Should be good to go.
